### DISCUSSION TITLE
Making a function pure by discarding logs

### DISCUSSION THREAD

**Discussion by qwbarch** [kowainik/co-log#224] [Q&A] [answered]
I have a function with a pure value that can log messages:
```haskell
{-# LANGUAGE FlexibleContexts #-}
{-# LANGUAGE OverloadedStrings #-}

import Colog (Message, WithLog, logInfo, logMessagePure, runPureLog, usingLoggerT)

zeroM :: WithLog r Message m => m Int
zeroM = 0 <$ logInfo "Hello world!"
```
I'd like to remove its monadic context by discarding logs. I found an example that shows how to do this: https://github.com/co-log/co-log/blob/72fbe394b437c698d574cd7604ad3f7f807383e0/co-log/tutorials/Main.hs#L204

Here is my attempt on discarding the log messages of ``zeroM`` to make it completely pure:
```haskell
zero :: Int
zero = fst . runPureLog $ usingLoggerT logMessagePure zeroM
```
However, the types does not line up and my compiler complains. How do I discard the logs, keeping only returned value?

**Comment by chshersh** [comment, marked as answer]
Hi @qwbarch 👋🏻 

The `logMessagePure` function is polymorphic over the message type:

```haskell
logMessagePure :: Monad m => LogAction (PureLoggerT msg m) msg
```

Because of how type inference works, Haskell is not able to infer that `msg` is the same as `Message` in your example. A simple solution would be to give Haskell hint by specifying the type of `logMessagePure` explicitly:

```haskell
zero :: Int
zero = fst . runPureLog $ usingLoggerT (logMessagePure :: LogAction (PureLogger Message) Message) zeroM
```

Alternatively, you can enable the `TypeApplications` extension and specify this type a bit shorter:

```
zero = fst . runPureLog $ usingLoggerT (logMessagePure @_ @Message) zeroM
```

Moreover, there's even a better way to ignore logging. `PureLogger` still performs logging and stores all logs in memory. This might be inefficient. If you don't care about them, you can just discard the logs instead:

```haskell
zero :: Int
zero = runIdentity $ usingLoggerT (mempty :: LogAction Identity Message) zeroM
```

**Reply by qwbarch** [reply]
Perfect, that last example is what I need. Thanks a lot!
